File: scripts/lib/jspace/retrieve.py
"""Discovery context assembly — how the agent 'thinks' about RE problems."""
from __future__ import annotations

from .embed import expand_re_query
from .store import JSpaceStore

DISCOVERY_PREAMBLE = """# RE discovery context (JSpace)
You are investigating a binary with Ghidra. Use the retrieved cards below as
working memory before calling MCP tools. Prefer evidence (addresses, names,
strings) over speculation. When unsure, decompile / list xrefs via Ghidra MCP.
"""


def search(store: JSpaceStore, query: str, *, top_k: int = 8, kind: str | None = None) -> list[dict]:
    expanded = expand_re_query(query)
    return store.hybrid_search(expanded, top_k=top_k, kind=kind)


def format_card(hit: dict, idx: int) -> str:
    name = hit.get("name") or "(unnamed)"
    addr = hit.get("address") or ""
    kind = hit.get("kind") or "chunk"
    score = hit.get("score", 0.0)
    via = hit.get("via") or ""
    text = (hit.get("text") or "")[:1200]
    return (
        f"### [{idx}] {kind} `{name}` @ {addr} (score={score:.3f} via={via})\n"
        f"{text}\n"
    )


def discovery_context(
    store: JSpaceStore,
    query: str,
    *,
    top_k: int = 8,
    current_function: str | None = None,
    current_decompile: str | None = None,
) -> str:
    """Build a mental-model pack for the agent: selection + JSpace neighbors."""
    parts = [DISCOVERY_PREAMBLE, f"## Question\n{query}\n"]
    if current_function:
        parts.append(f"## Current selection\nFunction: `{current_function}`\n")
        if current_decompile:
            parts.append("```c\n" + current_decompile[:2000] + "\n```\n")
    hits = search(store, query, top_k=top_k)
    parts.append("## Retrieved RE cards\n")
    if not hits:
        parts.append("_JSpace empty — run `ghidra-vibe-jspace index` after MCP is up._\n")
    else:
        for i, h in enumerate(hits, 1):
            parts.append(format_card(h, i))
    parts.append(
        "## Suggested next MCP moves\n"
        "- `methods` / select a card name\n"
        "- `decompile` the top hit\n"
        "- `xrefs` / `get_xrefs_to` for callers\n"
        "- rename once the role is clear\n"
    )
    return "\n".join(parts)

File: rust/ghidra-vibe-tools/src/retrieve.rs
use crate::embed::expand_re_query;
use crate::store::JSpaceStore;
use anyhow::Result;
use serde_json::Value;

const DISCOVERY_PREAMBLE: &str = "\
# RE discovery context (JSpace)
You are investigating a binary with Ghidra. Use the retrieved cards below as
working memory before calling MCP tools. Prefer evidence (addresses, names,
strings) over speculation. When unsure, decompile / list xrefs via Ghidra MCP.
";

pub fn search(
    store: &JSpaceStore,
    query: &str,
    top_k: usize,
    kind: Option<&str>,
) -> Result<Vec<Value>> {
    let expanded = expand_re_query(query);
    store.hybrid_search(&expanded, top_k, kind)
}

fn format_card(hit: &Value, idx: usize) -> String {
    let name = hit["name"].as_str().unwrap_or("(unnamed)");
    let addr = hit["address"].as_str().unwrap_or("");
    let kind = hit["kind"].as_str().unwrap_or("chunk");
    let score = hit["score"].as_f64().unwrap_or(0.0);
    let via = hit["via"].as_str().unwrap_or("");
    let text = hit["text"].as_str().unwrap_or("");
    let text: String = text.chars().take(1200).collect();
    format!(
        "### [{idx}] {kind} `{name}` @ {addr} (score={score:.3} via={via})\n{text}\n"
    )
}

pub fn discovery_context(
    store: &JSpaceStore,
    query: &str,
    top_k: usize,
    current_function: Option<&str>,
    current_decompile: Option<&str>,
) -> Result<String> {
    let mut parts = vec![
        DISCOVERY_PREAMBLE.to_string(),
        format!("## Question\n{query}\n"),
    ];
    if let Some(fn_name) = current_function {
        parts.push(format!("## Current selection\nFunction: `{fn_name}`\n"));
        if let Some(dec) = current_decompile {
            let slice: String = dec.chars().take(2000).collect();
            parts.push(format!("```c\n{slice}\n```\n"));
        }
    }
    let hits = search(store, query, top_k, None)?;
    parts.push("## Retrieved RE cards\n".into());
    if hits.is_empty() {
        parts.push(
            "_JSpace empty — run `ghidra-vibe-jspace index` after MCP is up._\n".into(),
        );
    } else {
        for (i, h) in hits.iter().enumerate() {
            parts.push(format_card(h, i + 1));
        }
    }
    parts.push(
        "## Suggested next MCP moves\n\
         - `methods` / select a card name\n\
         - `decompile` the top hit\n\
         - `xrefs` / `get_xrefs_to` for callers\n\
         - for Swift: confirm Demangler Swift + Type Metadata ran\n\
         - rename once the role is clear\n"
            .into(),
    );
    Ok(parts.join("\n"))
}
